File: Data Structures & Algorithms/combination-target-sum/submission-3.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res=[]
        n = len(nums)
        def backtracking(i:int,curr):
            if i >= n or sum(curr)> target:
                return
            if sum(curr)==target:
                res.append(curr.copy())
                return
         
            curr.append(nums[i])
            backtracking(i,curr)
            curr.pop()
            backtracking(i+1,curr)
            
        backtracking(0,[])
        return res
            

File: Data Structures & Algorithms/subsets/submission-3.cpp
class Solution {
        vector<vector<int>> res;

public:
    void backtracking(vector<int>& nums , int i,vector<int> curr){
        if (i>=nums.size()){
            res.push_back(curr);
            return;
        }
        curr.push_back(nums[i]);
        backtracking(nums,i+1,curr);
        curr.pop_back();
        backtracking(nums,i+1,curr);

        


    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> curr;
        backtracking(nums,0,curr);
        return res;
    }
};
